#463772 [C#テキスト]13章演習問題

kurosu-being opened the pull request:
第13 EntityFrameworkによるデータアクセス

データベース管理システムについて、Entiity FrameworkのCode Firstを利用しながら学習しました。

kurosu-being committed:
#463772 [C#テキスト]13章演習問題

Practice13-1-1~2を実装しました。

kurosu-being committed:
#463772 [C#テキスト]13章演習問題

Practice13-1-1~2において、インデントを修正しました。

kurosu-being committed:
#463772 [C#テキスト]13章演習問題

Practice13-1-3を実装しました。

kurosu-being committed:
#463772 [C#テキスト]13章演習問題

Practice13-1-3において、メソッド名を変更しました（GetBooksLongestTitle→EnumerateBooksLongestTitle）
Practice13-1-4のファイルを作成しました。

kurosu-being committed:
#463772 [C#テキスト]13章演習問題

Practice13-1-4を実装しました。

kurosu-being committed:
#463772 [C#テキスト]13章演習問題

Practice13-1-5を実装しました。
Practice13-1-1~2において、コメントを修正しました。
Practice13-1-1~2～Practice13-1-4において、インデントの修正を行いました。

odaira1119 commented on 章末問題13(修正)/Practice13-1-1~2/Models/BooksDbContext.cs:
```diff
@@ -0,0 +1,32 @@
+    //{
+    //    public int Id { get; set; }
+    //    public string Name { get; set; }
+    //}
```
コメントアウトしているコード（不要なコード）は、削除するようにしてください。

kurosu-being replied:
ご指摘ありがとうございます。
削除いたしました。

odaira1119 commented on 章末問題13(修正)/Practice13-1-1~2/Program13-1-1~13-1-2.cs:
```diff
@@ -0,0 +1,92 @@
+                Console.WriteLine($"{wBook.Title}{wBook.PublishedYear}");
+            }
+            Console.ReadLine();
+        }
```
DisplayAllBooksメソッドは、すべての書籍を表示するメソッドですが、
実際に中身を見ると、GetBooksメソッドを呼んでいるため、書籍を取得して表示するメソッドになっています。
なので、コメントを修正するか引数で書籍を受け取り、受け取った書籍を表示するメソッドに修正してください。

kurosu-being replied:
ご指摘ありがとうございます。
引数で書籍を受け取り、受け取った書籍を表示するメソッドとして修正しました。また、メソッド名をDisplayBooksと変更しました。

odaira1119 commented on 章末問題13(修正)/Practice13-1-1~2/Program13-1-1~13-1-2.cs:
```diff
@@ -0,0 +1,92 @@
+        /// </summary>
+        static void DisplayAllBooks() {
+            var wBooks = GetBooks();
+            foreach (var wBook in wBooks) {
```
wBooks は、一度しか使われていないようです。
一度しか使われない変数は、わざわざ変数で管理する必要がありません。
また、無駄にメモリを確保することになります。

kurosu-being replied:
ご指摘ありがとうございます。
不要な変数を削除いたしました。

odaira1119 commented on 章末問題13(修正)/Practice13-1-1~2/Program13-1-1~13-1-2.cs:
```diff
@@ -0,0 +1,92 @@
+                wDb.Authors.AddRange(wAuthorsToAdd);
+                wDb.SaveChanges();
+            }
+        }
```
AddAuthorsメソッドは、著者を追加するメソッドですが、
このメソッドを呼ぶたびにメソッド内で初期化された著者が追加されるようになっています。
そのため、追加したい著者を変えたい場合、メソッドの中身を書き換える必要があり、使うときにとても注意が必要となっています。
汎用的に好きな著者を追加できるメソッドに修正してみてください。

kurosu-being replied:
ご指摘ありがとうございます。
汎用的なものになるよう、リストを受け取り加えるメソッドになるよう修正し、
Mainメソッドでリストを追加するように書き換えました。

odaira1119 commented on 章末問題13(修正)/Practice13-1-1~2/Program13-1-1~13-1-2.cs:
```diff
@@ -0,0 +1,92 @@
+                wDb.Books.AddRange(wBooksToAdd);
+                wDb.SaveChanges();
+            }
+        }
```
こちらも決まった書籍を追加するメソッドになっていますので、
AddAuthorsメソッドと同様に好きな書籍を追加できるように修正してみてください。

kurosu-being replied:
ご指摘ありがとうございます。
同様に修正いたしました。

odaira1119 commented on 章末問題13(修正)/Practice13-1-3/Program13-1-3.cs:
```diff
@@ -0,0 +1,112 @@
+        /// <returns>最もタイトルの長い書籍</returns>
+        static IEnumerable<Book> EnumerateBooksLongestTitle(BooksDbContext vContext) {
+            var wMaxLength = vContext.Books.Max(x => x.Title.Length);
+            return vContext.Books.Where(x => x.Title.Length == wMaxLength).ToList();
```
ToList()が可読性を下げ、不要なメモリの確保を行っています。
削除するかメソッドの戻り値の型とメソッド名の修正、のどちらかを行ってください。

kurosu-being replied:
ご指摘ありがとうございます。
ToList()を削除いたしました。

odaira1119 commented on 章末問題13(修正)/Practice13-1-3/Program13-1-3.cs:
```diff
@@ -0,0 +1,112 @@
+            var wMaxLength = vContext.Books.Max(x => x.Title.Length);
+            return vContext.Books.Where(x => x.Title.Length == wMaxLength).ToList();
+        }
+    }
```
引数に関するコメントが抜けています。
また、BooksDbContext を引数で受け取るのは大きすぎると思います。
このメソッド内で必要な最小限のものを受け取るように修正してください。

kurosu-being replied:
ご指摘ありがとうございます。
引数に対するコメントを追記し、受け取る型をIEnumerable<Book>に変更いたしました。

odaira1119 commented on 章末問題13(修正)/Practice13-1-4/Program13-1-4.cs:
```diff
@@ -0,0 +1,117 @@
+        /// 発行年の古い順に指定した数の書籍を取得するメソッド
+        /// </summary>
+        static IEnumerable<Book> GetOldestThreeBooks(BooksDbContext vContext) {
+            return vContext.Books.OrderBy(x => x.PublishedYear).Take(3).ToList();
```
上記と同様にこちらのToList()も修正してください。

kurosu-being replied:
ご指摘ありがとうございます。
同様に修正いたしました。

odaira1119 commented on 章末問題13(修正)/Practice13-1-4/Program13-1-4.cs:
```diff
@@ -0,0 +1,117 @@
+        /// <returns>最もタイトルの長い書籍</returns>
+        static IEnumerable<Book> DisplayBooksWithLongestTitle(BooksDbContext vContext) {
+            var wMaxLength = vContext.Books.Max(x => x.Title.Length);
+            return vContext.Books.Where(x => x.Title.Length == wMaxLength).ToList();
```
上記と同様にこちらのToList()も修正してください。

kurosu-being replied:
ご指摘ありがとうございます。
こちらも同様に、ToList()を削除する形で修正しました。

odaira1119 commented on 章末問題13(修正)/Practice13-1-4/Program13-1-4.cs:
```diff
@@ -0,0 +1,117 @@
+        static IEnumerable<Book> DisplayBooksWithLongestTitle(BooksDbContext vContext) {
+            var wMaxLength = vContext.Books.Max(x => x.Title.Length);
+            return vContext.Books.Where(x => x.Title.Length == wMaxLength).ToList();
+        }
```
最小限の型を受け取るように引数を修正してください。

kurosu-being replied:
ご指摘ありがとうございます。
BooksDbContext →IEnumerable<Book>に型を変更致しました。

odaira1119 commented on 章末問題13(修正)/Practice13-1-4/Program13-1-4.cs:
```diff
@@ -0,0 +1,117 @@
+        static IEnumerable<Book> GetOldestThreeBooks(BooksDbContext vContext) {
+            return vContext.Books.OrderBy(x => x.PublishedYear).Take(3).ToList();
+        }
+    }
```
最小限の型を受け取るように引数を修正してください。

kurosu-being replied:
ご指摘ありがとうございます。
同様に、BooksDbContext →IEnumerableに型を変更致しました。

odaira1119 commented on 章末問題13(修正)/Practice13-1-5/Program13-1-5.cs:
```diff
@@ -0,0 +1,124 @@
+        /// </summary>
+        static IEnumerable<IGrouping<Author, Book>> GetBooksByAuthor(BooksDbContext vContext) {
+            return vContext.Books.Include("Author").ToList().GroupBy(x => x.Author).OrderByDescending(group => group.Key.Birthday);
+        }
```
実行してみたところ、group.Keyがnullとなりnull参照例外が起きました。
再度、考え直してみてください。
また、groupではなくyを使ってください。x,y,zの順に使われます。

kurosu-being replied:
ご指摘ありがとうございます。
group.Keyがnullの時（一度も書籍の追加を行っていないとき）「コレクションが存在しません」と表記されるよう
修正いたしました。
また、groupではなくyを使用して修正いたしました。

odaira1119 reviewed (changes requested):
コメントしましたので、確認してください。

また、BookとAuthorがコンストラクタを使用せずにオブジェクト初期化子を使っていて、冗長に感じます。
コンストラクタを定義し、コンストラクタを使って初期化するようにしてください。

kurosu-being committed:
#463772 [C#テキスト]13章演習問題

全ての問に対してコンストラクタを実装し、修正しました。

kurosu-being committed:
#463772 [C#テキスト]13章演習問題

Practice13-1-1~2において、コメントを修正しました。書籍と著者の追加メソッドが汎用的なものになるよう修正しました。

kurosu-being committed:
#463772 [C#テキスト]13章演習問題

Practice13-1-1~2において明示的な型を利用しました。

kurosu-being committed:
Revert "#463772 [C#テキスト]13章演習問題"

This reverts commit 5e495917a0f463ff3204001352e0207d7f990b36.

kurosu-being committed:
#463772 [C#テキスト]13章演習問題

Practice13-1-1~2において、コンストラクタを実装しました。
Practice13-1-3,13-1-4において、メソッドの引数を小さなものに変更しました。

kurosu-being committed:
#463772 [C#テキスト]13章演習問題

Practice13-1-1~2において、確認の為、一度Dbをクリアにするメソッドを追記しました。
各csファイルにおいて不要な参照を削除しました。
各csファイルにおいて、Practice13-1-1~2で実装したメソッドと同様になるように揃えました。
Practice13-1-5において、GroupByがnullであった時の処理を追記しました。

kurosu-being committed:
#463772 [C#テキスト]13章演習問題

Practice13-1-1~2において、クラスに対するコメントを追加しました。

odaira1119 commented on 章末問題13(修正)/Practice13-1-1~2/Program13-1-1~13-1-2.cs:
```diff
@@ -0,0 +1,113 @@
+        /// <summary>
+        /// 書籍を挿入するメソッド（本文で作成したメソッド）
+        /// </summary>
+        static void InsertBooks(BooksDbContext dbContext, List<Book> booksToAdd) {
```
コーディング規約（引数名：先頭にvを付ける。例：vAuthor）に沿って、書き換えてください。
他のメソッドも確認してください

kurosu-being replied:
ご指摘ありがとうございます。
修正いたしました。

odaira1119 commented on 章末問題13(修正)/Practice13-1-1~2/Program13-1-1~13-1-2.cs:
```diff
@@ -0,0 +1,113 @@
+        /// 書籍を表示するメソッド
+        /// </summary>
+        /// <param name="books">表示する書籍のコレクション</param>
+        static void DisplayBooks(IEnumerable<Book> books) {
```
引数名をコーディング規約に沿って修正してください。

kurosu-being replied:
ご指摘ありがとうございます。
引数名をコーディング規約に沿って修正いたしました。

odaira1119 commented on 章末問題13(修正)/Practice13-1-1~2/Program13-1-1~13-1-2.cs:
```diff
@@ -0,0 +1,113 @@
+        /// <summary>
+        /// 著者を追加するメソッド
+        /// </summary>
+        private static void AddAuthors(List<Author> authorsToAdd) {
```
引数名をコーディング規約に沿って修正してください。

kurosu-being replied:
ご指摘ありがとうございます。
引数名をコーディング規約に沿って修正いたしました。

odaira1119 commented on 章末問題13(修正)/Practice13-1-1~2/Program13-1-1~13-1-2.cs:
```diff
@@ -0,0 +1,113 @@
+        /// <summary>
+        /// 書籍を追加するメソッド
+        /// </summary>
+        private static void AddBooks(List<Book> booksToAdd) {
```
引数名をコーディング規約に沿って修正してください。

kurosu-being replied:
ご指摘ありがとうございます。
引数名をコーディング規約に沿って修正いたしました。

odaira1119 commented on 章末問題13(修正)/Practice13-1-1~2/Program13-1-1~13-1-2.cs:
```diff
@@ -0,0 +1,113 @@
+
+                // すべての著者を取得し削除
+                var wAllAuthors = wDbContext.Authors.ToList();
+                wDbContext.Authors.RemoveRange(wAllAuthors);
```
`wAllBooks`と`wAllAuthors`が一度しか使われていないので、変数でもつ必要がないと思います。
1度しか使われない変数は、必要な場合もありますが、基本的には宣言しません。
不要な変数を宣言しないように修正してください。

kurosu-being replied:
ご指摘ありがとうございます。
不要な変数を削除し、修正いたしました。

odaira1119 commented on 章末問題13(修正)/Practice13-1-3/Program13-1-3.cs:
```diff
@@ -0,0 +1,91 @@
+        /// <summary>
+        /// 書籍を挿入するメソッド（本文で作成したメソッド）
+        /// </summary>
+        static void InsertBooks(BooksDbContext dbContext, List<Book> booksToAdd) {
```
引数名をコーディング規約に沿って修正してください。

kurosu-being replied:
ご指摘ありがとうございます。
引数名をコーディング規約に沿って修正いたしました。

odaira1119 commented on 章末問題13(修正)/Practice13-1-3/Program13-1-3.cs:
```diff
@@ -0,0 +1,91 @@
+        /// 書籍を表示するメソッド
+        /// </summary>
+        /// <param name="books">表示する書籍のコレクション</param>
+        static void DisplayBooks(IEnumerable<Book> books) {
```
引数名をコーディング規約に沿って修正してください。

kurosu-being replied:
ご指摘ありがとうございます。
引数名をコーディング規約に沿って修正いたしました。

odaira1119 commented on 章末問題13(修正)/Practice13-1-3/Program13-1-3.cs:
```diff
@@ -0,0 +1,91 @@
+        /// <summary>
+        /// 著者を追加するメソッド
+        /// </summary>
+        private static void AddAuthors(List<Author> authorsToAdd) {
```
引数名をコーディング規約に沿って修正してください。

kurosu-being replied:
ご指摘ありがとうございます。
引数名をコーディング規約に沿って修正いたしました。

odaira1119 commented on 章末問題13(修正)/Practice13-1-3/Program13-1-3.cs:
```diff
@@ -0,0 +1,91 @@
+        /// <summary>
+        /// 書籍を追加するメソッド
+        /// </summary>
+        private static void AddBooks(List<Book> booksToAdd) {
```
引数名をコーディング規約に沿って修正してください。

kurosu-being replied:
ご指摘ありがとうございます。
引数名をコーディング規約に沿って修正いたしました。

odaira1119 commented on 章末問題13(修正)/Practice13-1-4/Program13-1-4.cs:
```diff
@@ -0,0 +1,96 @@
+        /// <summary>
+        /// 書籍を挿入するメソッド（本文で作成したメソッド）
+        /// </summary>
+        static void InsertBooks(BooksDbContext dbContext, List<Book> booksToAdd) {
```
引数名をコーディング規約に沿って修正してください。

kurosu-being replied:
ご指摘ありがとうございます。
引数名をコーディング規約に沿って修正いたしました。

odaira1119 commented on 章末問題13(修正)/Practice13-1-4/Program13-1-4.cs:
```diff
@@ -0,0 +1,96 @@
+        /// 書籍を表示するメソッド
+        /// </summary>
+        /// <param name="books">表示する書籍のコレクション</param>
+        static void DisplayBooks(IEnumerable<Book> books) {
```
引数名をコーディング規約に沿って修正してください。

kurosu-being replied:
ご指摘ありがとうございます。
引数名をコーディング規約に沿って修正いたしました。

odaira1119 commented on 章末問題13(修正)/Practice13-1-4/Program13-1-4.cs:
```diff
@@ -0,0 +1,96 @@
+        /// <summary>
+        /// 著者を追加するメソッド
+        /// </summary>
+        private static void AddAuthors(List<Author> authorsToAdd) {
```
引数名をコーディング規約に沿って修正してください。

kurosu-being replied:
ご指摘ありがとうございます。
引数名をコーディング規約に沿って修正いたしました。

odaira1119 commented on 章末問題13(修正)/Practice13-1-4/Program13-1-4.cs:
```diff
@@ -0,0 +1,96 @@
+        /// <summary>
+        /// 書籍を追加するメソッド
+        /// </summary>
+        private static void AddBooks(List<Book> booksToAdd) {
```
引数名をコーディング規約に沿って修正してください。

kurosu-being replied:
ご指摘ありがとうございます。
引数名をコーディング規約に沿って修正いたしました。

odaira1119 commented on 章末問題13(修正)/Practice13-1-5/Program13-1-5.cs:
```diff
@@ -0,0 +1,131 @@
+            var wBooksWithAuthor = vContext.Books
+                                    .Include("Author")
+                                    .Where(book => book.Author != null) 
+                                    .ToList();
```
ToList()で不要にメモリを確保しているので、削除しましょう。
同時にしたのif条件式も修正してください。

kurosu-being replied:
ご指摘ありがとうございます。
ToList()を削除し、次のif条件式をAny()メソッドで検索するように修正しました。

odaira1119 commented on 章末問題13(修正)/Practice13-1-5/Program13-1-5.cs:
```diff
@@ -0,0 +1,131 @@
+
+            return wBooksWithAuthor
+                           .GroupBy(book => book.Author)
+                           .OrderByDescending(group => group.Key.Birthday);
```
こちらもコーディング規約に沿って、xyzの順に使うようにしましょう

kurosu-being replied:
ご指摘ありがとうございます。
引数名をコーディング規約に沿って修正いたしました。

kurosu-being committed:
463772 [C#テキスト]13章演習問題

Practice13-1-1~2において、LINQの中の変数名を変更しました。

odaira1119 reviewed (commented):
再度、コメントしました。
コーディング規約に関する些細な指摘ばかりですが、修正お願いします。
コーディング規約に慣れていきましょう！

kurosu-being committed:
463772 [C#テキスト]13章演習問題

コーディング規約に基づき、引数名を変更しました。

kurosu-being committed:
#463772 [C#テキスト]13章演習問題

コーディング規約に基づき、引数名を変更しました。

odaira1119 commented on 章末問題13(修正)/Practice13-1-5/Program13-1-5.cs:
```diff
@@ -0,0 +1,131 @@
+                                    .Where(x => x.Author != null) ;
+                                    
+
+            if (wBooksWithAuthor.Any()) {
```
条件が逆転してしまいました。
正しい条件になるように修正お願いします。

kurosu-being replied:
失礼しました、修正を行いました。🙇‍♀️

odaira1119 reviewed (commented):
1点、修正お願いします。

kurosu-being committed:
463772 [C#テキスト]13章演習問題

Practice13-1-5において、ifの条件を修正しました。
